### ISSUE TITLE
Samtools 1.3.1 (using htslib 1.3.1) DoS with malformed BAM

### ISSUE THREAD

**Issue by JohnLonginotto** [samtools/samtools#632]
Bgzf blocks have an extended gzip attribute for compressed data (in bytes). This is usually 2 bytes to store a number as large as 65536 - the maximum size for a compressed bgzf block. However if you change the gzip header to allow for a number as large as 4294967296, and then compress more than 65536 bytes, this results in a valid gzip file - but will cause samtools to hang while using 99% CPU.

To test you can run [bambomb.py](https://github.com/JohnLonginotto/bambomb.py/blob/master/bambomb.py) with a command like

`./bambomb.py -t2 -r 2000 | samtools view`

Note that the number of compressed reads (-r) needs to be over a certain number such that the compressed data is over 65536 bytes. Lower values do not trigger the error. This doesn't effect versions of samtools <1.0, only more recent versions.

**Comment by jmarshall** [comment]
A file produced with `./bambomb.py -t2 -r 2000` has three GZIP member blocks.  The first one is an ordinary BGZF block and is read by _bgzf.c_ as such.  The second one has a **4-byte** BC extra field which does not match the BGZF spec, so is read by _bgzf.c_ as plain GZIP.  The hang then occurs [in `inflate_gzip_block()`](https://github.com/samtools/htslib/blob/1.3.2/bgzf.c#L451-L452) as `inflate()` is repeatedly called without making any progress, continually returning `Z_BUF_ERROR`.

Fixing this may be as simple as this _htslib/bgzf.c_ change:

```diff
@@ -508,8 +508,7 @@ static int inflate_gzip_block(BGZF *fp, int cached)
             fp->gz_stream->avail_out = BGZF_MAX_BLOCK_SIZE - fp->block_offset;
             fp->gz_stream->msg = NULL;
             ret = inflate(fp->gz_stream, Z_NO_FLUSH);
-            if ( ret==Z_BUF_ERROR ) continue;   // non-critical error
-            if ( ret<0 ) {
+            if (ret < 0 && ret != Z_BUF_ERROR) {
                 if (hts_verbose >= 1) {
                     fprintf(stderr, "[E::%s] inflate failed: %s\n",
                             __func__,

```

but really there is no such thing as a limited-size **gzip** block and this function deserves a careful audit.

**Comment by jkbonfield** [comment]
That fix looks logical, but agreed it needs more careful analysis.  When it comes to the issue of oversized XLEN subfields, this can't be as big as you claim as it is 16-bit maximum. 

Right now bgzf implementation is overly strict as it requires exactly 6 bytes of extra data consisting of BC tag only, although the spec implies BC field can be anywhere within a larger XLEN block (all following the 2 byte ID, 2 byte len and BSIZE data as per RFC1952).  If we relaxed this we'd probably need to be checking against BGZF_MAX_BLOCK_SIZE (or just add 64k to it) so we have room for a full block with header.

**Comment by jmarshall** [comment]
> the implementation is overly strict as it requires exactly 6 bytes of extra data consisting of BC tag only

This is an area where the spec does not reflect reality, in the opposite way to usual.  While the description of BGZF allows for other GZIP extra data subfields, for all practical purposes none exist anywhere except for BGZF's `BC` and both major implementations have always required exactly six bytes here.  I would not be surprised if other implementations did as well.

See also the related discussion in samtools/htslib#451.  So rather than relax this, IMHO if action is to be taken it should be to tighten the spec to reflect reality in BGZF's GZIP extra data subfields.

**Comment by jkbonfield** [comment]
Agreed, it was just an "if we" comment.  However thinking about it even without bgzf format we probably run into the same issue of buffer size with pure gzip parsing.

I've constructed a test block with 64k XLEN and will be prodding it to see what chokes.

**Comment by jkbonfield** [comment]
Commit https://github.com/samtools/htslib/commit/d8d0323bca21cfd131e2f183d12c9dd31a5ed75b fixes this and several other oddities with raw gzip streams (ie not bgzf), incorporating @jmarshall 's suggestion above.  I don't seem able to get any deadlocks or buffer overruns with my hacked up gzip streams, but please let us know if your fuzzer finds any more issues. 

I'm leaving this open for a little longer for you to test.

Thanks.

**Comment by JohnLonginotto** [comment]
Thank you James (and John!) for the help - and can i just say how very impressed I am at how quick the samtools team interact and fix user concerns, really fantastic support!

The fuzzer checks all the things I could think of so it is unlikely I will find any other issues. I'm not a security guy so this is not really my department. Of course, if anything else does come up i'll make you are aware, but it is unlikely :)

All the best,
